site-src/geps/gep-1016.md

```diff
@@ -0,0 +1,813 @@
+- [the device plugin framework](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/)
+
+Given gRPC's importance in the application-layer networking space and to
+the Kubernetes project in particular, we must ensure that the gRPC control plane
+configuration landscape does not Balkanize.
+
+### Encapsulated Network Protocols
+
+At the time of writing, the only kind of route officially defined by the Gateway
+APIs is `HttpRoute`. This GEP is novel not only in that it introduces a second
+protocol to route, but also in that it introduces the first protocol
+encapsulated in a protocol already supported by the API.
+
+That is, it _is_ theoretically possible to route gRPC traffic using only `HTTPRoute`
+resources, but there are several serious problems with forcing gRPC users to route traffic at
+the level of HTTP. This is why we propose a new resource.
+
+In setting this precendent, we must also introduce a coherent policy for _when_
+to introduce a custom `Route` resource for an encapsulated protocol for which a
+lower layer protocol already exists. We propose the following criteria for such
+an addition.
+
+- Users of the encapsulated protocol would miss out on significant conventional features from their ecosystem if forced to route at a lower layer.
+- Users of the enapsulated protocol would experience a degraded user experience if forced to route at a lower layer.
+- The encapsulated protocol has a significant user base, particularly in the Kubernetes community.
+
+gRPC meets _all_ of these criteria and is therefore, we contend, a strong
+candidate for inclusion in the Gateway APIs.
+
+#### HTTP/2 Cleartext
+
+gRPC allows HTTP/2 cleartext communication (H2C). This is conventionally deployed for
+testing. Many control plane implementations do not support this by default and
+would require special configuration to work properly.
+
+#### Content-Based Routing
+
+While not included in the scope of this initial GEP, a common use case cited for
+routing gRPC is payload-aware routing. That is, routing rules which determine a
+backend based on the contents of the protocol buffer payload.
+
+#### User Experience
+
+The user experience would also degrade significantly if forced to route at the level of HTTP.
+
+- Encoding services and methods as URIs (an implementation detail of gRPC)
+- The [Transfer Encoding header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Transfer-Encoding) for trailers
+- Many features supported by HTTP/2 but not by gRPC, such as
+  - Query parameters
+  - Methods besides `POST`
+  - CORS
+
+
+#### Proxyless Service Mesh
+
+The gRPC library supports proxyless service mesh, a system by which routing
+configuration is received not by an in-line proxy or sidecar proxy but by the client
+itself. Eventually, `GRPCRoute` in the Gateway APIs should support this feature.
+However, to date, there are no HTTP client libraries capable of participating
+in a proxyless service mesh.
```

Comment:
I'm not very familiar with proxy-less SM. What are the responsibilities of implementation to support this? What changes (if any) are required in this API to support this?

Reply:
Let me start by caveating that proxyless service mesh support is not actually part of *this* proposal.

With that said, I don't believe any changes will be necessary to the `GRPCRoute` resource to enable this. Instead, attachment to a `Mesh` object, referenced through the`parentRefs` field will activate this support. Another proposal for a `Mesh` resource is planned for sometime in the next couple of quarters.